assembly/src/parsers/ast/serialization.rs

```diff
@@ -0,0 +1,1165 @@
+        *bytes = remaining_bytes;
+        let result = usize::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
+
+    pub fn get_string(bytes: &mut &[u8]) -> Result<String, SerializationError> {
+        let (mut length, remaining_bytes) = bytes.split_at(8);
+        let length = ByteReader::get_usize(&mut length).unwrap();
+        let (string, remaining_bytes) = remaining_bytes.split_at(length);
+        *bytes = remaining_bytes;
+        let string = String::from_utf8(string.to_vec()).unwrap();
+        Ok(string)
+    }
+
+    pub fn get_felt(bytes: &mut &[u8]) -> Result<Felt, SerializationError> {
+        Ok(Felt::new(ByteReader::get_u64(bytes).unwrap()))
+    }
+
+    pub fn get_instruction(bytes: &mut &[u8]) -> Result<Instruction, SerializationError> {
+        Deserializable::read_from(bytes)
+    }
+
+    pub fn get_node(bytes: &mut &[u8]) -> Result<Node, SerializationError> {
+        Deserializable::read_from(bytes)
+    }
+
+    pub fn get_vec_node(bytes: &mut &[u8]) -> Result<Vec<Node>, SerializationError> {
+        Deserializable::read_from(bytes)
+    }
+
+    pub fn get_procedure(bytes: &mut &[u8]) -> Result<ProcedureAst, SerializationError> {
+        ProcedureAst::read_from(bytes)
+    }
+}
+
+pub trait Deserializable: Sized {
+    fn read_from(bytes: &mut &[u8]) -> Result<Self, SerializationError>;
+}
+
+impl Deserializable for Vec<Node> {
+    fn read_from(bytes: &mut &[u8]) -> Result<Self, SerializationError> {
+        let mut vec_node: Vec<Node> = Vec::new();
+        let vec_len = ByteReader::get_u16(bytes).unwrap();
+
+        for _ in 0..vec_len {
+            vec_node.push(ByteReader::get_node(bytes).unwrap());
+        }
+
+        Ok(vec_node)
+    }
+}
+
+impl Deserializable for Node {
+    fn read_from(bytes: &mut &[u8]) -> Result<Self, SerializationError> {
+        let first_byte = ByteReader::read_u8(bytes).unwrap();
+
+        match first_byte {
+            253 => {
+                ByteReader::get_u8(bytes).unwrap();
+                Ok(Node::IfElse(
+                    ByteReader::get_vec_node(bytes).unwrap(),
+                    ByteReader::get_vec_node(bytes).unwrap(),
+                ))
+            }
+            254 => {
+                ByteReader::get_u8(bytes).unwrap();
+                Ok(Node::Repeat(
+                    ByteReader::get_usize(bytes).unwrap(),
+                    ByteReader::get_vec_node(bytes).unwrap(),
+                ))
+            }
+            255 => {
+                ByteReader::get_u8(bytes).unwrap();
+                Ok(Node::While(ByteReader::get_vec_node(bytes).unwrap()))
+            }
+            _ => Ok(Node::Instruction(
+                ByteReader::get_instruction(bytes).unwrap(),
+            )),
+        }
+    }
+}
+
+impl Deserializable for Instruction {
+    fn read_from(bytes: &mut &[u8]) -> Result<Self, SerializationError> {
+        let code = ByteReader::get_u8(bytes).unwrap();
+
+        match code {
```

Comment:
If we have `read_opcode()` implemented as mentioned above, this `match` would match against `OpCode` enum.